shotover/src/codec/opensearch.rs

```diff
@@ -0,0 +1,332 @@
+
+        let mut builder = Response::builder().status(response.code.unwrap());
+
+        let builder_headers = builder.headers_mut().unwrap();
+
+        for header in response.headers {
+            if header.name.is_empty() && header.value.is_empty() {
+                break;
+            }
+            builder_headers.insert(
+                HeaderName::from_bytes(header.name.as_bytes()).unwrap(),
+                HeaderValue::from_bytes(header.value).unwrap(),
+            );
+        }
+
+        let r = builder.body(()).unwrap();
+        let content_length = match r.headers().get(header::CONTENT_LENGTH) {
+            Some(cl) => match atoi::atoi(cl.as_bytes()) {
+                Some(cl) => cl,
+                None => return Err(anyhow!("invalid content-length header")),
+            },
+            None => return Err(anyhow!("missing content-length header")), // content-length header is required in HTTP responses
+        };
+        let (parts, _) = r.into_parts();
+        Ok(Some((
+            body_start,
+            HttpHead::Response(ResponseParts {
+                version: parts.version,
+                status: parts.status,
+                headers: parts.headers,
+            }),
+            content_length,
+        )))
+    }
+}
+
+#[derive(Debug)]
+enum State {
+    ParsingResponse,
+    ReadingBody(HttpHead, usize),
+}
+
+impl Decoder for OpenSearchDecoder {
+    type Item = Messages;
+    type Error = CodecReadError;
+
+    fn decode(&mut self, src: &mut BytesMut) -> Result<Option<Self::Item>, CodecReadError> {
+        if src.is_empty() {
+            return Ok(None);
+        }
+
+        loop {
+            match std::mem::replace(&mut self.state, State::ParsingResponse) {
+                State::ParsingResponse => {
+                    let decode_result = if self.direction == Direction::Source {
+                        self.decode_request(src).map_err(CodecReadError::Parser)?
+                    } else {
+                        self.decode_response(src).map_err(CodecReadError::Parser)?
+                    };
+
+                    if let Some((body_start, parts, content_length)) = decode_result {
+                        self.state = State::ReadingBody(parts, content_length);
+                        src.advance(body_start);
+                    } else {
+                        return Ok(None);
+                    };
+                }
+                State::ReadingBody(parts, content_length) => {
+                    if src.is_empty() {
+                        // sometimes messages will have a content-length but no
+                        // body ?
+                        return Ok(Some(vec![Message::from_frame(Frame::OpenSearch(
+                            OpenSearchFrame::new(parts, bytes::Bytes::new()),
+                        ))]));
+                    }
+
+                    if src.len() < content_length {
+                        self.state = State::ReadingBody(parts, content_length);
```

Comment:
this looks like a noop to me. 
If we reach this line then we are already in the ReadingBody state.